src/k8s_sandbox/_kubernetes_api.py

```diff
@@ -2,43 +2,157 @@
     return namespace
 
 
-def _ensure_config_loaded() -> None:
-    with _load_config_lock:
-        global _config_loaded
-        if not _config_loaded:
-            config.load_kube_config()
-            _config_loaded = True
+def get_current_context_name() -> str:
+    """Get the name of the current kubeconfig context.
+
+    As defined by the kubeconfig file.
+    """
+    context = _Config.get_instance().get_context(None)
+    return context["name"]
+
+
+def validate_context_name(context_name: str) -> None:
+    """Validate that the current kubeconfig context is a valid context.
+
+    If the context is invalid, a ValueError is raised.
+    """
+    _Config.get_instance().get_context(context_name)
+
+
+class _Config:
+    """A thread-safe singleton that holds a snapshot of the kubeconfig file.
+
+    Loaded only once for performance and thread-safety.
+
+    This config can become out of date if the underlying kubeconfig file on disk is
+    modified after it is loaded.
+    """
+
+    _load_lock: threading.Lock = threading.Lock()
+    _instance: _Config | None = None
+
+    def __init__(
+        self,
+        contexts: list[dict[str, Any]] | None,
+        current_context: dict[str, Any] | None,
+    ):
+        self.contexts = contexts
+        self.current_context = current_context
+
+    @classmethod
+    def get_instance(cls) -> _Config:
+        # The kubernetes.config module is not thread-safe.
+        with cls._load_lock:
+            if cls._instance is None:
+                # Must call load_kube_config() before any clients created with the
+                # current context can be used.
+                config.load_kube_config()
+                cls._instance = _Config(*config.list_kube_config_contexts())
+        return cls._instance
+
+    @classmethod
+    def ensure_loaded(cls) -> None:
+        cls.get_instance()
+
+    def get_context(self, context_name: str | None) -> dict[str, Any]:
+        if context_name is None:
+            return self._get_current_context()
+        return self._get_named_context(context_name)
+
+    def _get_current_context(self) -> dict:
+        if self.current_context is None:
+            raise ValueError(
+                "Could not get the current context because the current context is not "
+                "set in the kubeconfig file."
+            )
+        return self.current_context
+
+    def _get_named_context(self, context_name: str) -> dict:
+        if not self.contexts:
+            raise ValueError(
+                f"Could not find a context named '{context_name}' in kubeconfig "
+                "because no contexts were present in the kubeconfig file."
+            )
+        for context in self.contexts:
+            if context["name"] == context_name:
+                return context
+        raise ValueError(
+            f"Could not find a context named '{context_name}' in the kubeconfig file. "
+            f"Available contexts: {self.contexts}."
```

Comment:
Should this maybe print a list of names? I assume this will be a lot of output otherwise.
Something like this? (not tested!)
```suggestion
            f"Available contexts: {",".join([context["name"] for context in self.contexts])}."
```

Reply:
That would be much more helpful! Actioned. Sample error:
```
ValueError("Could not find a context named 'does-not-exist' in the kubeconfig file. Available contexts: ['arn:aws:eks:eu-west-2:302263072849:cluster/rpv2-prod', 'arn:aws:eks:us-east-1:302263072849:cluster/rpv2-prod', 'minikube'].")
```